File: metacasa-ios/MetaCasa/Models/Bill.swift
import Foundation

/// Vencimiento (bill) — una obligación de pago con fecha específica.
/// Port de `BillForm` / `BillCard` del web (App.jsx:653-800).
struct Bill: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var title: String
    var description: String?
    var amount: Decimal
    var currency: String
    var dueDate: Date
    var status: BillStatus
    var paidAt: Date?
    var category: String?
    var accountId: UUID?
    var note: String?
    var recurring: Bool
    let createdBy: UUID
    let createdAt: Date?
    var updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case title
        case description
        case amount
        case currency
        case dueDate = "due_date"
        case status
        case paidAt = "paid_at"
        case category
        case accountId = "account_id"
        case note
        case recurring
        case createdBy = "created_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    /// Estado visual según `dueDate` y `status`.
    /// Refleja los colores del web (App.jsx:5706-5724).
    enum UrgencyLevel {
        case paid
        case overdue        // pasada, no pagada
        case dueToday       // hoy
        case dueSoon        // 1-3 días
        case upcoming       // 4-14 días
        case future         // >14 días
        case skipped
    }

    var urgency: UrgencyLevel {
        if status == .paid { return .paid }
        if status == .skipped { return .skipped }
        let cal = Calendar.current
        let today = cal.startOfDay(for: Date())
        let due = cal.startOfDay(for: dueDate)
        let days = cal.dateComponents([.day], from: today, to: due).day ?? 0
        if days < 0 { return .overdue }
        if days == 0 { return .dueToday }
        if days <= 3 { return .dueSoon }
        if days <= 14 { return .upcoming }
        return .future
    }

    var daysUntilDue: Int {
        let cal = Calendar.current
        let today = cal.startOfDay(for: Date())
        let due = cal.startOfDay(for: dueDate)
        return cal.dateComponents([.day], from: today, to: due).day ?? 0
    }
}

enum BillStatus: String, Codable, Hashable, Sendable {
    case pending, paid, skipped
    var label: String {
        switch self {
        case .pending: return "Pendiente"
        case .paid:    return "Pagado"
        case .skipped: return "Saltado"
        }
    }
}

File: metacasa-flutter/lib/models/bill.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';

part 'bill.freezed.dart';
part 'bill.g.dart';

/// Estado del vencimiento.
enum BillStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('paid')
  paid,
  @JsonValue('skipped')
  skipped;

  String get label => switch (this) {
        BillStatus.pending => 'Pendiente',
        BillStatus.paid => 'Pagado',
        BillStatus.skipped => 'Saltado',
      };
}

/// Estado visual según `dueDate` y `status`. Refleja los colores del web.
enum BillUrgencyLevel {
  paid,
  overdue, // pasada, no pagada
  dueToday, // hoy
  dueSoon, // 1-3 días
  upcoming, // 4-14 días
  future, // >14 días
  skipped,
}

/// Vencimiento (bill) — obligación de pago con fecha específica.
///
/// Espejo 1:1 de la tabla REAL `public.bills` (verificada contra la DB viva,
/// que está más actualizada que el modelo Swift de iOS). Campos del schema:
/// id, user_id, household_id, title, amount, currency, due_date, status,
/// recurrence_type, reminder_days, amount_original, currency_original,
/// fx_rate_to_base, category, created_at.
@freezed
abstract class Bill with _$Bill {
  const Bill._();

  const factory Bill({
    required String id,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'household_id') required String householdId,
    required String title,
    @DecimalConverter() required Decimal amount,
    @Default('USD') String currency,
    @JsonKey(name: 'due_date') required DateTime dueDate,
    @Default(BillStatus.pending) BillStatus status,
    @Default('') String category,
    @JsonKey(name: 'recurrence_type') String? recurrenceType,
    @JsonKey(name: 'reminder_days') int? reminderDays,
    @JsonKey(name: 'amount_original')
    @DecimalNullConverter()
    Decimal? amountOriginal,
    @JsonKey(name: 'currency_original') String? currencyOriginal,
    @JsonKey(name: 'fx_rate_to_base')
    @DecimalNullConverter()
    Decimal? fxRateToBase,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _Bill;

  factory Bill.fromJson(Map<String, dynamic> json) => _$BillFromJson(json);

  /// `true` si el vencimiento se repite (tiene `recurrence_type` definido).
  bool get isRecurring => recurrenceType != null && recurrenceType!.isNotEmpty;

  /// Días hasta el vencimiento (negativo si ya pasó). Normalizado a inicio
  /// de día, como en iOS.
  int get daysUntilDue {
    final DateTime now = DateTime.now();
    final DateTime today = DateTime(now.year, now.month, now.day);
    final DateTime due = DateTime(dueDate.year, dueDate.month, dueDate.day);
    return due.difference(today).inDays;
  }

  /// Nivel de urgencia visual. Port de `Bill.urgency` en iOS.
  BillUrgencyLevel get urgency {
    if (status == BillStatus.paid) return BillUrgencyLevel.paid;
    if (status == BillStatus.skipped) return BillUrgencyLevel.skipped;
    final int days = daysUntilDue;
    if (days < 0) return BillUrgencyLevel.overdue;
    if (days == 0) return BillUrgencyLevel.dueToday;
    if (days <= 3) return BillUrgencyLevel.dueSoon;
    if (days <= 14) return BillUrgencyLevel.upcoming;
    return BillUrgencyLevel.future;
  }
}
